sale_order_type/models/sale_order.py

```diff
@@ -1,56 +1,53 @@
-# -*- encoding: utf-8 -*-
-##############################################################################
-# For copyright and license notices, see __openerp__.py file in root directory
-##############################################################################
+# -*- coding: utf-8 -*-
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
-from openerp import api, fields, models
+from odoo import api, fields, models
 
 
 class SaleOrder(models.Model):
     _inherit = 'sale.order'
 
     def _get_order_type(self):
-        return self.env['sale.order.type'].search([])[:1]
+        return self.env['sale.order.type'].search([], limit=1)
 
     type_id = fields.Many2one(
         comodel_name='sale.order.type', string='Type', default=_get_order_type)
 
     @api.multi
-    def onchange_partner_id(self, part):
-        res = super(SaleOrder, self).onchange_partner_id(part)
-        if part:
-            partner = self.env['res.partner'].browse(part)
-            res['value'].update({
-                'type_id': partner.sale_type.id or self._get_order_type().id,
-            })
-        return res
+    @api.onchange('partner_id')
+    def onchange_partner_id(self):
+        super(SaleOrder, self).onchange_partner_id()
+        if self.partner_id:
+            self.type_id = self.partner_id.sale_type or self._get_order_type()
```

Comment:
self._get_order_type() is called by default. It shouldn't be recalled at the onchange_partner